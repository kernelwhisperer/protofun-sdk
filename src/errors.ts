export class IndexerError extends Error {
  name = "IndexerError"
}
