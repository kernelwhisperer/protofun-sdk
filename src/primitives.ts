export type Timeframe = "Block" | "Minute" | "Hour" | "Day" | "Week"

export enum PriceUnit {
  ETH = "ETH",
  USD = "USD",
  GWEI = "Gwei",
}

export type Block = {
  baseFeePerGas: string // BigInt
  burnedFees: string // BigInt
  firstGasPrice: string // BigInt
  gasFees: string // BigInt
  gasUsed: string // BigInt
  id: string // bytes
  lastGasPrice: string // BigInt
  maxGasPrice: string // BigInt
  minGasPrice: string // BigInt
  minerTips: string // BigInt
  number: string // BigInt
  timestamp: string // BigInt
  // txns: Array<Txn>
  txnCount: number
}

export type SimpleBlock = Omit<Block, "txns" | "timestamp"> & {
  timestamp: string
}

export type Candle = {
  close: string
  high: string
  low: string
  open: string
  timestamp: string
}

export function isCandle(value: unknown): value is Candle {
  return typeof value === "object" && value !== null && "open" in value
}

export function isCandleArray(value: unknown[]): value is Candle[] {
  return Array.isArray(value) && value.length > 0 && isCandle(value[0])
}

export type QueryRequest = {
  /**
   * @default 1000
   */
  limit?: number
  priceUnit?: PriceUnit
  since?: string
  timeframe: Timeframe
  until?: string
}
export type QueryResult = Promise<Candle[] | SimpleBlock[]>
export type QueryFn = (request: QueryRequest) => QueryResult
