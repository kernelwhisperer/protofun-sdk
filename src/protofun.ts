export * from "./errors"
export * from "./metric-declarations"
export * from "./primitives"
