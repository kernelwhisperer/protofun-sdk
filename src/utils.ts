import { MetricId, ProtocolId, QueryFn, SubscribeFn, Timeframe } from "./protofun"

export async function wait(ms: number): Promise<void> {
  return new Promise((resolve) => setTimeout(resolve, ms))
}

export const allTimeframes: Timeframe[] = ["Block", "Minute", "Hour", "Day", "Week"]

export type MetricFnsResult = {
  query: QueryFn
  subscribe: SubscribeFn
  supportedTimeframes: Timeframe[]
}

export async function loadMetricFns(
  protocolId: ProtocolId,
  metricId: MetricId,
  packagePath = "protofun/dist/metrics"
): Promise<MetricFnsResult> {
  const module = await import(`${packagePath}/${protocolId}/${metricId}`)
  const { default: query, supportedTimeframes = allTimeframes, subscribe } = module
  return { query, subscribe, supportedTimeframes }
}

export function getLowestTimeframe(supportedTimeframes: Timeframe[]) {
  return supportedTimeframes.find((x) => x !== "Block") as Timeframe
}
