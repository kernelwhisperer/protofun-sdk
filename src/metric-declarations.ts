import { PriceUnit, Timeframe } from "./primitives"
import { allTimeframes } from "./utils"

const METRIC_DECLARATIONS = {
  aave: [],
  btc: ["btc_price"],
  comp: ["tvl", "comp_price"],
  eth: ["base_fee", "eth_price", "tx_cost"],
  mkr: [],
} as const

/**
 * Protocols
 */
export type ProtocolId = keyof typeof METRIC_DECLARATIONS
export type Protocol = {
  enabled?: boolean
  id: ProtocolId
  title: string
  wip?: boolean
}

export const PROTOCOL_MAP: Record<ProtocolId, Protocol> = {
  aave: {
    id: "aave",
    title: "Aave",
  },
  btc: {
    enabled: true,
    id: "btc",
    title: "Bitcoin",
  },
  comp: {
    enabled: true,
    id: "comp",
    title: "Compound",
    wip: true,
  },
  eth: {
    enabled: true,
    id: "eth",
    title: "Ethereum",
  },
  mkr: {
    id: "mkr",
    title: "MakerDAO",
  },
} as const

export const PROTOCOL_IDS = Object.keys(PROTOCOL_MAP) as ProtocolId[]

export const PROTOCOLS = Object.values(PROTOCOL_MAP).sort((a, b) => {
  if (a.enabled === b.enabled) {
    return 0
  } else if (a.enabled && !b.enabled) {
    return -1
  } else {
    return 1
  }
})

/**
 * Metrics
 */

// Not working:
// type ValueOf<T> = T[keyof T];
// export type MetricId = ValueOf<typeof METRIC_LIST_DEFINITION>[number];
export type MetricIdForProtocol<T extends ProtocolId> = typeof METRIC_DECLARATIONS[T][number]
export type MetricId = MetricIdForProtocol<ProtocolId>

export type Variant = {
  label: string
  precision: number
  value?: string
}

export type Metric = {
  allowCompactPriceScale?: boolean
  disallowAlerts?: boolean
  disallowCandleType?: boolean
  disallowLiveMode?: boolean
  id: MetricId
  precision: number
  preferredLogScale?: boolean
  priceUnits: PriceUnit[]
  protocol: ProtocolId
  significantDigits: number[]
  timeframes: Timeframe[]
  title: string
  variants?: Variant[]
  wip?: boolean
}

// type MapType<T extends ProtocolId> = Record<
//   T,
//   Record<MetricIdsOfProtocol<T>, Metric>
// >;
type MetricsMapType = Record<ProtocolId, Partial<Record<MetricId, Metric>>>

export const METRICS_MAP: MetricsMapType = {
  aave: {},
  btc: {
    btc_price: {
      id: "btc_price",
      precision: 1,
      preferredLogScale: true,
      priceUnits: [PriceUnit.USD],
      protocol: "btc",
      significantDigits: [0],
      timeframes: ["Minute", "Hour", "Day", "Week"],
      title: "BTC price",
    },
  },
  comp: {
    comp_price: {
      id: "comp_price",
      precision: 1,
      preferredLogScale: true,
      priceUnits: [PriceUnit.USD],
      protocol: "comp",
      significantDigits: [2],
      timeframes: ["Minute", "Hour", "Day", "Week"],
      title: "Comp price",
    },
    tvl: {
      allowCompactPriceScale: true,
      disallowCandleType: true,
      id: "tvl",
      precision: 1,
      priceUnits: [PriceUnit.USD],
      protocol: "comp",
      significantDigits: [0],
      timeframes: ["Hour", "Day"],
      title: "Total value locked",
      variants: [{ label: "V3 USDC - Wrapped Ether Pool", precision: 1 }],
      wip: true,
    },
  },
  eth: {
    base_fee: {
      id: "base_fee",
      precision: 1e9,
      preferredLogScale: true,
      priceUnits: [PriceUnit.GWEI],
      protocol: "eth",
      significantDigits: [2],
      timeframes: allTimeframes,
      title: "Base fee per gas",
    },
    eth_price: {
      id: "eth_price",
      precision: 1,
      preferredLogScale: true,
      priceUnits: [PriceUnit.USD],
      protocol: "eth",
      significantDigits: [2],
      timeframes: ["Minute", "Hour", "Day", "Week"],
      title: "Ether price",
    },
    tx_cost: {
      id: "tx_cost",
      precision: 1,
      preferredLogScale: true,
      priceUnits: [PriceUnit.USD, PriceUnit.ETH],
      protocol: "eth",
      significantDigits: [2, 5],
      timeframes: ["Minute", "Hour", "Day", "Week"],
      title: "Transaction cost",
      variants: [
        { label: "ETH Transfer", precision: 1e18 / 21_000 },
        { label: "ERC20 Approval", precision: 1e18 / 45_000 },
        { label: "ERC20 Transfer", precision: 1e18 / 65_000 },
        { label: "NFT Transfer", precision: 1e18 / 85_000 },
        { label: "Uniswap V2 Swap", precision: 1e18 / 150_000 },
        { label: "Uniswap V3 Swap", precision: 1e18 / 185_000 },
        { label: "L2 Deposit", precision: 1e18 / 250_000 },
      ],
    },
  },
  mkr: {},
}

export const METRICS = PROTOCOL_IDS.map((protocolId) =>
  Object.values(METRICS_MAP[protocolId] || {})
).flat()

export function getMetric(protocolId: ProtocolId, metricId: MetricId): Metric {
  const metric = METRICS_MAP[protocolId][metricId]

  if (!metric) {
    throw new Error("Metric not found.")
  }

  return metric
}

export function getMetricPrecision(metric: Metric, variantIndex: number): number {
  const { variants, precision: defaultPrecision } = metric
  return variants ? variants[variantIndex].precision : defaultPrecision
}

export function getSignificantDigits(metric: Metric, priceUnitIndex: number): number {
  return metric.significantDigits[priceUnitIndex]
}
