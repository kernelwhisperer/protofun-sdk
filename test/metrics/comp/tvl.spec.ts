import { expect, it } from "vitest"

import query, { subscribe } from "../../../src/metrics/comp/tvl"
import { wait } from "../../../src/protofun"

it("Daily candles", async () => {
  // act
  const candles = await query({
    timeframe: "Day",
    // since: "0",
    until: "1661811308",
  })
  // assert
  expect(candles).toMatchSnapshot()
})

it("Hourly candles", async () => {
  // act
  const candles = await query({
    timeframe: "Hour",
    // since: "0",
    until: "1661494432",
  })
  // assert
  expect(candles).toMatchSnapshot()
})

it("Weekly candles", async () => {
  // act
  try {
    await query({
      timeframe: "Week",
      // since: "0",
      until: "1661811308",
    })
  } catch (error) {
    // assert
    expect(error).toMatchInlineSnapshot(
      "[Error: Timeframe 'Week' is not supported for this metric.]"
    )
  }
})

it("Subscribe", async () => {
  // arrange
  const timeframe = "Hour"
  const candles = await query({
    limit: 1,
    timeframe,
  })
  // act
  const unsubscribe = subscribe({
    onNewData: (data) => {
      candles.push(data)
      unsubscribe()
    },
    pollingInterval: 1000,
    since: candles[0].timestamp,
    timeframe,
  })
  // assert
  await wait(3_000)
  expect(candles.length).to.equal(2)
})
