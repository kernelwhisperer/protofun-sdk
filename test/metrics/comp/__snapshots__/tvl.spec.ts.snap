// Vitest Snapshot v1

exports[`Daily candles 1`] = `
[
  {
    "close": "0",
    "high": "0",
    "low": "0",
    "open": "0",
    "timestamp": "1660369110",
  },
  {
    "close": "22450.5300709537031295429801111",
    "high": "22450.5300709537031295429801111",
    "low": "22450.5300709537031295429801111",
    "open": "22450.5300709537031295429801111",
    "timestamp": "1661552003",
  },
  {
    "close": "1385222.64918099069034989163723976",
    "high": "1385222.64918099069034989163723976",
    "low": "1385222.64918099069034989163723976",
    "open": "1385222.64918099069034989163723976",
    "timestamp": "1661635342",
  },
  {
    "close": "3546739.70069824994089027657905939",
    "high": "3546739.70069824994089027657905939",
    "low": "3546739.70069824994089027657905939",
    "open": "3546739.70069824994089027657905939",
    "timestamp": "1661721824",
  },
  {
    "close": "3661658.7094895565372932501535048",
    "high": "3661658.7094895565372932501535048",
    "low": "3661658.7094895565372932501535048",
    "open": "3661658.7094895565372932501535048",
    "timestamp": "1661811308",
  },
]
`;

exports[`Hourly candles 1`] = `
[
  {
    "close": "0",
    "high": "0",
    "low": "0",
    "open": "0",
    "timestamp": "1660369110",
  },
  {
    "close": "3721.153284837",
    "high": "3721.153284837",
    "low": "3721.153284837",
    "open": "3721.153284837",
    "timestamp": "1661478127",
  },
  {
    "close": "6115.1706",
    "high": "6115.1706",
    "low": "6115.1706",
    "open": "6115.1706",
    "timestamp": "1661482395",
  },
  {
    "close": "4670.53269273",
    "high": "4670.53269273",
    "low": "4670.53269273",
    "open": "4670.53269273",
    "timestamp": "1661489242",
  },
  {
    "close": "6377.49126157",
    "high": "6377.49126157",
    "low": "6377.49126157",
    "open": "6377.49126157",
    "timestamp": "1661494432",
  },
]
`;
